File: examples/format/inmemory-go/store.go
// Package main — rat-format-inmemory-go: a reference `kind: format` plugin.
//
// This is a sub-phase-0d reference implementation (ADR-003): its job is to
// exercise + validate the format/v1 wire contract, not to be production storage.
// It keeps tables in memory as ordered row maps. "Bulk data" that the contract
// says flows out-of-band as Arrow is, for this reference, carried inline through
// a trivial in-process stream registry (see stream.go) — the point is to prove
// the CONTROL-plane contract (Resolve/Append/Merge/Overwrite/Maintain), with a
// stand-in for the data leg, so a second independent impl + golden vectors can
// cross-check behavior.
package main

import (
	"sort"
	"sync"
)

// row is one record: a map of column -> value. Values are strings to keep the
// reference trivial and language-neutral (the real Arrow handoff carries typed
// columns; this stand-in does not need types to validate the control contract).
type row map[string]string

// table is an in-memory dataset addressed by a TableRef.identifier.
type table struct {
	rows     []row
	snapshot int64 // bumped on every mutation; surfaced as WriteResult.snapshot_id
}

// store is the plugin's whole state: identifier -> table. Concurrency-safe so the
// gRPC server can serve parallel calls.
type store struct {
	mu     sync.Mutex
	tables map[string]*table
}

func newStore() *store {
	return &store{tables: map[string]*table{}}
}

func (s *store) get(identifier string) *table {
	t, ok := s.tables[identifier]
	if !ok {
		t = &table{}
		s.tables[identifier] = t
	}
	return t
}

// append adds rows unconditionally. Returns rows affected + new snapshot id.
func (s *store) append(identifier string, rows []row) (int64, int64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	t := s.get(identifier)
	t.rows = append(t.rows, rows...)
	t.snapshot++
	return int64(len(rows)), t.snapshot
}

// overwrite replaces all rows. Returns rows affected + new snapshot id.
func (s *store) overwrite(identifier string, rows []row) (int64, int64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	t := s.get(identifier)
	t.rows = append([]row(nil), rows...)
	t.snapshot++
	return int64(len(rows)), t.snapshot
}

// merge upserts rows matched on mergeKeys: a source row replaces the first
// existing row whose mergeKey columns all match, else it is appended. Returns
// rows affected (updated + inserted) + new snapshot id.
func (s *store) merge(identifier string, mergeKeys []string, src []row) (int64, int64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	t := s.get(identifier)
	var affected int64
	for _, sr := range src {
		idx := -1
		for i, existing := range t.rows {
			if keysMatch(existing, sr, mergeKeys) {
				idx = i
				break
			}
		}
		if idx >= 0 {
			t.rows[idx] = sr
		} else {
			t.rows = append(t.rows, sr)
		}
		affected++
	}
	t.snapshot++
	return affected, t.snapshot
}

// scan returns a copy of the table's rows, optionally projecting columns. The
// rows are returned in a stable order (insertion order is preserved) so golden
// vectors are deterministic.
func (s *store) scan(identifier string, columns []string) []row {
	s.mu.Lock()
	defer s.mu.Unlock()
	t := s.get(identifier)
	out := make([]row, 0, len(t.rows))
	for _, r := range t.rows {
		if len(columns) == 0 {
			cp := make(row, len(r))
			for k, v := range r {
				cp[k] = v
			}
			out = append(out, cp)
			continue
		}
		cp := make(row, len(columns))
		for _, c := range columns {
			if v, ok := r[c]; ok {
				cp[c] = v
			}
		}
		out = append(out, cp)
	}
	return out
}

// maintain is a no-op upkeep for the in-memory store (nothing to compact). It
// still bumps the snapshot so the contract's "returns a WriteResult" holds.
func (s *store) maintain(identifier string) int64 {
	s.mu.Lock()
	defer s.mu.Unlock()
	t := s.get(identifier)
	t.snapshot++
	return t.snapshot
}

func keysMatch(a, b row, keys []string) bool {
	for _, k := range keys {
		if a[k] != b[k] {
			return false
		}
	}
	return true
}

// sortedKeys is a small helper for deterministic column ordering in tests.
func sortedKeys(r row) []string {
	ks := make([]string, 0, len(r))
	for k := range r {
		ks = append(ks, k)
	}
	sort.Strings(ks)
	return ks
}

File: plugins/format/inmemory-py/store.py
"""In-memory table store for rat-format-inmemory-py — the SECOND independent
`kind: format` reference (ADR-003).

Deliberately a from-scratch code path, not a port of the Go reference's internals:
its job is to prove the format/v1 wire contract is implementable by a different
implementation in a different language, and that it passes the SAME golden vectors
(contracts/conformance/format-v1.json). Tables are ordered lists of string-valued
rows; "bulk" data is carried by the in-process stream registry (streams.py), same
control-contract-only scope as the Go reference.
"""

import threading
from typing import Dict, List

Row = Dict[str, str]


class _Table:
    def __init__(self) -> None:
        self.rows: List[Row] = []
        self.snapshot: int = 0  # bumped on every mutation -> WriteResult.snapshot_id


class Store:
    """identifier -> _Table. Lock-guarded so the gRPC threadpool can serve in
    parallel."""

    def __init__(self) -> None:
        self._tables: Dict[str, _Table] = {}
        self._lock = threading.Lock()

    def _get(self, identifier: str) -> _Table:
        t = self._tables.get(identifier)
        if t is None:
            t = _Table()
            self._tables[identifier] = t
        return t

    def append(self, identifier: str, rows: List[Row]) -> tuple[int, int]:
        """Add rows unconditionally. Returns (rows_affected, new_snapshot)."""
        with self._lock:
            t = self._get(identifier)
            t.rows.extend(rows)
            t.snapshot += 1
            return len(rows), t.snapshot

    def overwrite(self, identifier: str, rows: List[Row]) -> tuple[int, int]:
        """Replace all rows. Returns (rows_affected, new_snapshot)."""
        with self._lock:
            t = self._get(identifier)
            t.rows = list(rows)
            t.snapshot += 1
            return len(rows), t.snapshot

    def merge(self, identifier: str, merge_keys: List[str], src: List[Row]) -> tuple[int, int]:
        """Upsert each source row onto the first existing row whose merge_keys all
        match, else append. Returns (rows_affected, new_snapshot)."""
        with self._lock:
            t = self._get(identifier)
            affected = 0
            for sr in src:
                idx = next((i for i, ex in enumerate(t.rows) if _keys_match(ex, sr, merge_keys)), -1)
                if idx >= 0:
                    t.rows[idx] = sr
                else:
                    t.rows.append(sr)
                affected += 1
            t.snapshot += 1
            return affected, t.snapshot

    def scan(self, identifier: str, columns: List[str]) -> List[Row]:
        """Return a copy of the rows in insertion order, optionally projecting
        columns (deterministic for golden vectors)."""
        with self._lock:
            t = self._get(identifier)
            if not columns:
                return [dict(r) for r in t.rows]
            return [{c: r[c] for c in columns if c in r} for r in t.rows]

    def maintain(self, identifier: str) -> int:
        """No-op upkeep (nothing to compact in memory); still bumps the snapshot so
        the contract's "returns a WriteResult" holds. Returns new_snapshot."""
        with self._lock:
            t = self._get(identifier)
            t.snapshot += 1
            return t.snapshot


def _keys_match(a: Row, b: Row, keys: List[str]) -> bool:
    return all(a.get(k) == b.get(k) for k in keys)
